File: Leetcode_Code/236. Lowest Common Ancestor of a Binary Tree/236.cpp
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        lowestCommonAncestor_dfs(root, p, q);
        return result;
    }

    private boolean lowestCommonAncestor_dfs(TreeNode root, TreeNode p, TreeNode q) {
        if (!terminate) {
            if (root == null) return false;
            boolean bl = lowestCommonAncestor_dfs(root.left, p, q);
            boolean br = lowestCommonAncestor_dfs(root.right, p, q);
            boolean now = root == p || root == q;
            if ((bl && br) || (bl && now) || (br && now)) {
                result = root;
                terminate = true;
            }
            return now || bl || br;
        }
        return false;
    }

    private TreeNode result = null;
    private boolean terminate = false;
}

File: Leetcode_Code/236. Lowest Common Ancestor of a Binary Tree/236.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def __init__(self):
        self.result = None
        self.terminate = False

    def lowestCommonAncestor(self, root, p, q):
        """
        :type root: TreeNode
        :type p: TreeNode
        :type q: TreeNode
        :rtype: TreeNode
        """
        if root is None:
            return None
        self.lowestCommonAncestor_dfs(root, p, q)
        return self.result

    def lowestCommonAncestor_dfs(self, root, p, q):
        if self.terminate == False:
            if root is None:
                return False
            bl = self.lowestCommonAncestor_dfs(root.left, p, q)
            br = self.lowestCommonAncestor_dfs(root.right, p, q)
            now = root == p or root == q
            if (bl and br) or (bl and now) or (br and now):
                self.result = root
                self.terminate = True
            return now or bl or br
